src/sentry/tasks/embeddings_grouping/utils.py

```diff
@@ -126,46 +126,56 @@ def initialize_backfill(
     else:
         last_processed_project_index_ret = last_processed_project_index
 
-    return project, last_processed_group_index_ret, last_processed_project_index_ret
+    return project, last_processed_group_id_ret, last_processed_project_index_ret
 
 
 @sentry_sdk.tracing.trace
-def get_current_batch_groups_from_postgres(project, last_processed_group_index, batch_size):
-    groups_to_backfill_query = (
+def get_current_batch_groups_from_postgres(project, last_processed_group_id, batch_size):
+    group_id_filter = Q()
+    if last_processed_group_id is not None:
+        group_id_filter = Q(id__lt=last_processed_group_id)
+
+    groups_to_backfill_batch = (
         Group.objects.filter(
+            group_id_filter,
             project_id=project.id,
             type=ErrorGroupType.type_id,
             times_seen__gt=1,
-            last_seen__gt=(datetime.now(UTC) - timedelta(days=90)),
         )
-        .exclude(status__in=[GroupStatus.PENDING_DELETION, GroupStatus.DELETION_IN_PROGRESS])
-        .values_list("id", "data")
-        .order_by("-times_seen", "id")
+        .values_list("id", "data", "status", "last_seen")
+        .order_by("-id")[:batch_size]
     )
-    total_groups_to_backfill_length = groups_to_backfill_query.count()
+    # Filter out groups that are pending deletion in memory so postgres won't make a bad query plan
+    groups_to_backfill_batch = [
+        (group[0], group[1])
+        for group in groups_to_backfill_batch
+        if group[2] not in [GroupStatus.PENDING_DELETION, GroupStatus.DELETION_IN_PROGRESS]
+        and group[3] > datetime.now(UTC) - timedelta(days=90)
+    ]
 
-    batch_end_index = min(last_processed_group_index + batch_size, total_groups_to_backfill_length)
-    groups_to_backfill_batch = groups_to_backfill_query[last_processed_group_index:batch_end_index]
+    total_groups_to_backfill_length = len(groups_to_backfill_batch)
+    batch_end_group_id = (
+        groups_to_backfill_batch[total_groups_to_backfill_length - 1][0]
+        if total_groups_to_backfill_length
+        else None
+    )
```

Comment:
It might be a good idea to get this id before you filter out rows in memory, just in case you happen to fetch a bunch of things that end up filtered out.

Reply:
good catch, thanks!